### ISSUE TITLE
@Context / @Info

### ISSUE THREAD

**Issue by phillip-kruger** [microprofile/microprofile-graphql#127]
Will still need to define how `@Context` or `@Info` will work.

From the meeting minutes:

- Add Environment interface to be injected by @Info 
- Check ResolutionEnvironment in SPQR, DataFetchingEnvironemt in GraphQL-Java
- See what we can leverage from existing APIs (like CDI)
- @Info should allow users to get the raw JSON for more complex (de-)serialization

**Comment by phillip-kruger** [comment]
Also see discussion in #167 , specifically point 4 from @andimarek

> Mapping the DataFetchingEnvironment and the possible return values is a big challenge: for example in order to use Batched loading efficiently you have to return CompletableFuture. Also returning Errors and Data is a very valid use case. See https://github.com/graphql-java/graphql-java/blob/master/src/main/java/graphql/execution/DataFetcherResult.java. There are also numerous other things inside DataFetchingEnvironment which are essential for writing production ready resolves.

**Comment by t1** [comment]
Defining all the types for a full data model would massively inflate the spec. Maybe we could start with a simpler mechanism and allow the injection of some implementation specific model, e.g. from graphql-java. Something like:

```java
@Inject
graphql.execution.ExecutionStepInfo info;
```

Using this would be non-portable, of course, but we could see what people actually use and define our own, reduced, portable model from there. Complex libraries, e.g., to dynamically build DB queries based on the GraphQL query, would probably build their own anti-corruption-layer, which would make it cheaper for them to go to a new model; and it would be a refined model to be suggested for the standard.

**Comment by phillip-kruger** [comment]
We did discuss allowing leaky abstraction. So it would rather be something like this:

```
@Inject
Context context; // From MP

ExecutionStepInfo esi = context.unwrap(); // Leak.
```

**Comment by beikov** [comment]
I am using the `DataFetchingEnvironment` in the graphql-java integration for Blaze-Persistence to support generating queries that do exactly what is need and I really like that you can access all of the request information. To do that, I need the selection list i.e. the properties that are requested, the requested type(inferred through the schema) and the arguments passed to the query. I hope this helps you.

For me, the `DataFetchingEnvironment` is perfectly fine as it gives all the information available. IMO MP-GraphQL should do the same and expose all available information.

**Comment by phillip-kruger** [comment]
Hi @beikov - Thanks for the feedback. This is exactly what I am working on at the moment. I hope to have the initial feature ready soon and would love to get your feedback then.

**Comment by t1** [comment]
Maybe `ExecutionStepInfo esi = context.unwrap(ExecutionStepInfo.class); // Leak.`?

**Comment by phillip-kruger** [comment]
No the unwrap would be a DataFetchingEnvironment. You can get the execution step info from there

**Comment by t1** [comment]
But the `DataFetchingEnvironment` is an interface from graphql-java. It shouldn't be directly in the spec, should it?

**Comment by phillip-kruger** [comment]
Yup, that is what I think we should leak in the unwrap

**Comment by t1** [comment]
So more like `ExecutionStepInfo esi = context.unwrap(DataFetchingEnvironment.class).getExecutionStepInfo();`

**Comment by phillip-kruger** [comment]
Something like that yes :)

**Comment by t1** [comment]
What would be in the `Context` interface other than the `unwrap` method? Providing a producer for the `DataFetchingEnvironment` would be possible without a change to the spec. It's a pure implementation detail.

**Comment by phillip-kruger** [comment]
I'll do a PR soon, then you can see... :)

**Comment by phillip-kruger** [comment]
Hi @t1, @beikov, @andymc12  and anyone else interested, please see https://github.com/smallrye/smallrye-graphql/pull/311/files#diff-04eaa96ad3729019e06e19590e848637 and comment there. 
Thanks :)

**Comment by beikov** [comment]
I guess the selection list part is ok, but what I really need is a way to build types programmatically and also introspect that metamodel then. `Context` should then give access to the resolved metamodel elements rather than strings.

I really like what graphql-java offers regarding `graphql.schema.idl.TypeDefinitionRegistry` which is pretty flexible. Is there a reason not to offer something similar?

Since GraphQL has no generics, I need to create synthetic types. Think of e.g. the Relay specifcation which more or less requires a Node, Edge and Conneciton type per user type. I wouldn't want a user to force implementing interfaces or require him to create explicit types to be able to use the Relay model.

**Comment by phillip-kruger** [comment]
Ok so you are saying rather than the field name, you need all meta data (from the schema) for that field ?

**Comment by beikov** [comment]
Right, but the main focus is on *having* a metamodel such that I can determine e.g. the return type of a query and being able to build GraphQL types programmatically to workaround the lack of generics support.

**Comment by marceloverdijk** [comment]
@beikov interesting thought about the synthetic types. That's my current pain as well to create these Node, Edge and Connection type per user type. Could you share more how you plan to create these synthetic types programmatically?

**Comment by beikov** [comment]
I am already doing that with graphql-java. You can see how I do it here: https://github.com/Blazebit/blaze-persistence/blob/master/integration/graphql/src/main/java/com/blazebit/persistence/integration/graphql/GraphQLEntityViewSupportFactory.java#L232

**Comment by phillip-kruger** [comment]
If you say "The return type of type of a query", do you mean the Java Object that then method annotated with `@Query` or `@Mutation` return ? And/Or every field's Java Object type ?

**Comment by beikov** [comment]
I mean the GraphQL Metamodel type from which the Java type should be accessible.

**Comment by phillip-kruger** [comment]
We do not have a GraphQL Metamodel.... what would you like to see in there ? Can you help with an example ?

**Comment by beikov** [comment]
Ideally I'd like to be able to access all information that can be modeled with GraphQL. In my use case specifically, I'd like to be able to determine the GraphQL type name and directives based on an initial type and a path. If I retrieve the return type of the currently executing query, I want to know the return type and from that return type, I want to be able to retrieve the type of a path e.g. `a/b` on which I access directives. The same should be possible with the selection list.

The string approach to the selection list is also not enough since one can use directives and arguments there as well. I think graphql-java modeled this pretty nicely with `graphql.schema.DataFetchingFieldSelectionSet`.

**Comment by phillip-kruger** [comment]
Hi @beikov. 
Thanks for the reply.

I think for now you can do this by using the `unwrap` and fall down to graphql-java, that gives us time to properly structure a model. I'll open a separate issue for that. Once we have a model we can make sure the model is available via injection. So the model does not have to be carried with the context (as the model is static) but you can look up things in the model from information in the context. Something like that. (We might even allow that in SmallRye in the mean time, as we have a model - I'll see if I can add that - that should give us an idea what should be in the model in the spec)

I think for phase 1 of this feature, we need to be able to do a proper JPA / SQL Query with the information. Then we can look at your use case one we have a model. (Since you should be able to still do what you want using `unwrap`)

@andymc12 - any thoughts ?

**Comment by beikov** [comment]
Of course I can do what I need to with the graphql-java API, I was just hoping that I could integrate with mp-graphql as well to be able to make it work for Quarkus, but I guess that will take some time.

If you are interested, Blaze-Persistence is capable of querying exactly what is requested on top of JPA. You define the model i.e. what should be accessible via Entity-Views for which GraphQL types are created by the integration.
The nice thing is, that it supports the Relay specification and uses keyset pagination under the hood with almost no work necessary.
Here some code: https://github.com/Blazebit/blaze-persistence/blob/master/examples/spring-data-graphql/src/main/java/com/blazebit/persistence/examples/spring/data/graphql/GraphQLProvider.java#L85

**Comment by phillip-kruger** [comment]
Thanks, I'll have a look. 

You should still be able to do this in Quarkus, because you will be able to use the unwrap to get to the graphql-java classes. (That is why we are adding the leaky abstraction, so that we can first determine usage before we move it to the spec)

**Comment by andimarek** [comment]
GraphQL Java maintainer here: @phillip-kruger If I look at your `Context` implementation of `getSelectedFields`  I am wondering what the motivation for that is: the whole `DataFetchingEnvironment` is available via `unwrap`, why didn't you delegate to it like you did with other methods? You are already using GraphQL Java types all over the place (e.g. GraphQLType) so it doesn't seem to me that there is a real layer between MP and GraphQL Java anyway. But maybe I am missing something, of course. Thanks!

**Comment by phillip-kruger** [comment]
Hi @andimarek Thanks for joining the discussion !! 
So what you see in the PR is all in SmallRye at the moment, and SmallRye use graphql-java to implement the specification. We use everything in the API module in smallrye to "bake" things that will eventually move to the spec. So that module should stay clean.

You will see there, in the API, we do not to include implementation details, but allowing the unwrap to happen until the API is mature enough. Users will use the `@Context` interface (that will move to the spec) and not the SR implementation directly. So even if in SmallRye, graphql-java is all over the place, users will have to consciously fall down to implementation details by using unwrap.

What would be cool is if we can create a serializable model that is separate from the spec and graphql-java, that just represent the schema, and then we both use it. Or something like that :)

Hope all this makes sense :)

**Comment by phillip-kruger** [comment]
Maybe one more thing to try and explain this. The same why that there is no Hibernate classes available in the JPA spec, we should not have any GraphQL classes in the MP spec.

**Comment by beikov** [comment]
So SmallRye GraphQL is an implementation of the MP-GraphQL spec based on graphql-java? That's nice! Then I only need the possibility to contribute GraphQL types to the `graphql.schema.idl.TypeDefinitionRegistry` programmatically to make this work. Any idea if/how I could do that?

**Comment by andimarek** [comment]
@phillip-kruger Thanks for explaining. This means you goal is to basically end up with something like JPA for GraphQL in Java. That correct?

**Comment by phillip-kruger** [comment]
Yes! The goal of MicroProfile GraphQL is that yes. A specification for code first GraphQL in Java. (so the JPA spec part) and SmallRye is an implementation of this spec. SPQR is another implementation (see https://github.com/leangen/graphql-spqr-modules/tree/master/graphql-spqr-module-microprofile) - so once the Context move up to MicroProfile Spec level, SPQR will also implement this.

**Comment by phillip-kruger** [comment]
@beikov to contribute to GraphQL types (`graphql.schema.idl.TypeDefinitionRegistry`) you need to chat to @andimarek :)

**Comment by andimarek** [comment]
SPQR is also GraphQL Java based, so I would assume it will look similar. But thanks for making the goal clear @phillip-kruger 

@beikov Anything specific to GraphQL Java you should open an issue at https://github.com/graphql-java/graphql-java or ask a question at https://spectrum.chat/graphql-java

**Comment by beikov** [comment]
Thanks @andimarek and @phillip-kruger but I would like to contribute types to `graphql.schema.idl.TypeDefinitionRegistry` in the context of MP-GraphQL, but AFAIU there is no way to hook into the bootstrap process of SmallRye GraphQL, or am I not seeing it? It would be great if I could receive a CDI event e.g. `@Observes AfterGraphQLTypeDiscovery e` which exposes something like the `graphql.schema.idl.TypeDefinitionRegistry` to be able to contribute types.

**Comment by phillip-kruger** [comment]
Hi @beikov - 
So you want to add to model types, but not using the annotations in MP ? We have discussed this in the spec meetings, there are 2 things here:

1) we are working on "adding annotations" to classes that you do not have control over. (see https://github.com/eclipse/microprofile-graphql/issues/111). But I do not think this is what you want, as you want to create this types dynamically right ?

2) Having the ability to do dynamic types (see https://github.com/eclipse/microprofile-graphql/issues/186). 

Both is still very much a WIP.

I am not sure I understand your requirement. But, what I can do in SmallRye (and then also Quarkus) is allow you to get an CDI Event just before we build the graphql-java types, so you can add to the model (see https://github.com/smallrye/smallrye-graphql/tree/master/common/schema-model) so that you can add your types there ? That might work ?

We build the above model from the annotations during build time and pass that to runtime, then we build the executable endpoint with graphql-java. Hope that makes sense. Let me know.

**Comment by beikov** [comment]
I see. Yeah I need to build partner types(Connection, Edge, Node) for each user type to be able to implement the necessary bits for e.g. the Relay specification.

>I am not sure I understand your requirement. But, what I can do in SmallRye (and then also Quarkus) is allow you to get an CDI Event just before we build the graphql-java types, so you can add to the model (see https://github.com/smallrye/smallrye-graphql/tree/master/common/schema-model) so that you can add your types there ? That might work ?

A hook at runtime to add additional types would be perfect. This is a necessary evil to support generic types :/

**Comment by phillip-kruger** [comment]
Cool, let me make that available in smallrye and then you can check if that works for you. I'll get back to you here once it's done.

**Comment by phillip-kruger** [comment]
@beikov - So I added this to SmallRye (not committed yet). You can anywhere in your code do this:

`public GraphQLSchema.Builder addDescription(@Observes GraphQLSchema.Builder builder) {

        GraphQLEnumType myOwnEnum = GraphQLEnumType.newEnum()
                .name("SomeEnum")
                .description("Adding some enum type")
                .value("value1")
                .value("value2").build();
        
        return builder.additionalType(myOwnEnum);
    }` 

As you can see you then have access to the builder just before we call build, so you can add additionalTypes etc.

Let see if this works for you. 

Note: This is SmallRye and graphql-java specific.

**Comment by beikov** [comment]
I don't have a problem with this being SmallRye and graphql-java specific, but I can't directly query existing types this way, I'd have to build a schema first before being able to do this. Also, I have to wire up the models manually.

Can't you expose the `TypeDefinitionRegistry` instead where I can use `graphql.language.TypeDefinition` objects which are then wired up automatically during schema building?

It should be possible to use a type in a query or mutation method which is defined by my extension. I don't know if at this point it's already too late, but with the `TypeDefinitionRegistry` approach, this should be perfectly fine.

**Comment by phillip-kruger** [comment]
Ok, let's create a issue in SmallRye for this. I am not sure I follow what you need. But it's outside of the scope of this issue. 
Will you do that ? https://github.com/smallrye/smallrye-graphql/issues

**Comment by mnguetsa** [comment]
Hello @phillip-kruger Just checking whether there has been some progress on this matter.
DataFetchingEnvironent access could be interesting ti differentiate between omitted parameters Vs actual null parameters.
Also I am open to alternative approaches to achieve this with Smallrye GraphQL
Thank you

**Comment by mnguetsa** [comment]
> You can do this in SmallRye GraphQL. Not sure if / when it will move to the spec

Thanks it worked
